service/controller/clusterapi/v16/resource_set.go

```diff
@@ -0,0 +1,104 @@
+package v16
+
+import (
+	"context"
+
+	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/controller"
+	"github.com/giantswarm/operatorkit/controller/resource/metricsresource"
+	"github.com/giantswarm/operatorkit/controller/resource/retryresource"
+	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+
+	"github.com/giantswarm/cluster-operator/pkg/cluster"
+	"github.com/giantswarm/cluster-operator/service/controller/clusterapi/v16/resources/awsclusterconfig"
+)
+
+// ResourceSetConfig contains necessary dependencies and settings for
+// Cluster API's Cluster controller ResourceSet configuration.
+type ResourceSetConfig struct {
+	BaseClusterConfig *cluster.Config
+	CMAClient         clientset.Interface
+	G8sClient         versioned.Interface
+	Logger            micrologger.Logger
+}
+
+// NewResourceSet returns a configured Cluster API's Cluster controller ResourceSet.
+func NewResourceSet(config ResourceSetConfig) (*controller.ResourceSet, error) {
+	var err error
+
+	if config.CMAClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.CMAClient must not be empty")
+	}
+	if config.G8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.CMAClient must not be empty")
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty")
+	}
+
+	var awsclusterconfigResource controller.Resource
+	{
+		c := awsclusterconfig.Config{
```

Comment:
I was thinking that for now it could be `awsclusterconfig` resources and later when we get other providers, we can try to extract common parts out into generic resources with configuration.